packages/server/src/threads/automation.ts

```diff
@@ -544,13 +290,288 @@ class Orchestrator {
+            status: AutomationErrors.MAX_ITERATIONS,
+            success: true,
+          }
+        )
+        shouldCleanup = false
+        break
+      }
+
+      let isFailure = false
+      const currentItem = this.getCurrentLoopItem(loopStep, loopStepIndex)
+      if (currentItem && typeof currentItem === "object") {
+        isFailure = Object.keys(currentItem).some(value => {
+          return currentItem[value] === loopStep?.inputs.failure
+        })
+      } else {
+        isFailure = currentItem && currentItem === loopStep.inputs.failure
+      }
+
+      if (isFailure) {
+        this.updateContextAndOutput(
+          loopStepIndex,
+          steps[stepToLoopIndex],
+          {
+            items: this.loopStepOutputs,
+            iterations: loopStepIndex,
+          },
+          {
+            status: AutomationErrors.FAILURE_CONDITION,
+            success: false,
+          }
+        )
+        shouldCleanup = false
+        break
+      }
+
+      this.context.steps[currentIndex + 1] = {
+        currentItem: this.getCurrentLoopItem(loopStep, loopStepIndex),
+      }
+
+      stepToLoopIndex = currentIndex + 1
+
+      await this.executeStep(steps[stepToLoopIndex], stepToLoopIndex)
+      iterationCount++
+    }
+
+    if (shouldCleanup) {
+      let tempOutput =
+        iterations === 0
+          ? {
+              status: AutomationStepStatus.NO_ITERATIONS,
+              success: true,
+            }
+          : {
+              success: true,
+              items: this.loopStepOutputs,
+              iterations: iterationCount,
+            }
+
+      // Loop Step clean up
+      this.executionOutput.steps.splice(currentIndex + 1, 0, {
+        id: steps[stepToLoopIndex].id,
+        stepId: steps[stepToLoopIndex].stepId,
+        outputs: tempOutput,
+        inputs: steps[stepToLoopIndex].inputs,
+      })
+      this.context.steps[currentIndex + 1] = tempOutput
+      this.loopStepOutputs = []
+    }
+
+    return stepToLoopIndex + 1
+  }
+  private async executeBranchStep(branchStep: BranchStep): Promise<void> {
+    const { branches, children } = branchStep.inputs
+
+    for (const branch of branches) {
+      const condition = await this.evaluateBranchCondition(branch.condition)
+      if (condition) {
+        const branchSteps = children?.[branch.name] || []
+        await this.executeSteps(branchSteps)
+        break
+      }
+    }
+  }
+
+  private async evaluateBranchCondition(
+    conditions: SearchFilters
+  ): Promise<boolean> {
+    const toFilter: Record<string, any> = {}
+
+    const processFilter = (filter: SearchFilters) => {
+      Object.entries(filter).forEach(([_, value]) => {
+        Object.entries(value).forEach(([field, _]) => {
+          const fromContext = processStringSync(
+            `{{ literal ${field} }}`,
+            this.context
+          )
+          toFilter[field] = fromContext
+        })
+      })
+      return filter
+    }
+    processFilter(conditions)
```

Comment:
Why are both calls required here? Should the recursing of the logical operators not automatically process filters at the top of the condition?

Reply:
In the `recurseLogicalOperators` function it only runs the passed function if it sees a logical operator, so I run the function on my top level first then we can do any recursive stuff after that.

Reply:
ah - I see, we should probably have a utility function which is just `recurseSearchFilters` which does both jobs in one!

Reply:
Done!